src/store/Controller.cc

```diff
@@ -354,19 +378,38 @@ Store::Controller::accumulateMore(StoreEntry &entry) const
 void
 Store::Controller::markForUnlink(StoreEntry &e)
 {
-    if (transients && e.mem_obj && e.mem_obj->xitTable.index >= 0)
+    if (transients)
         transients->markForUnlink(e);
-    if (memStore && e.mem_obj && e.mem_obj->memCache.index >= 0)
+    if (memStore)
         memStore->markForUnlink(e);
-    if (swapDir && e.swap_filen >= 0)
+    if (swapDir)
         swapDir->markForUnlink(e);
 }
 
 void
+Store::Controller::unlinkByKeyIfFound(const cache_key *key)
+{
+    if (StoreEntry *entry = static_cast<StoreEntry*>(hash_lookup(store_table, key))) {
+        debugs(20, 3, "got in-transit entry: " << *entry);
+        entry->release();
```

Comment:
> AFAICT, collapsed but still detached entries in other workers will not get stuck because StoreEntry::release() eventually calls Transients::abandon() that calls CollapsedForwarding::Broadcast(). 

While experimenting, I saw that this does not work properly with our changes. Currently, 'DELETE'  causes purgeEntriesByUrl(), Controller::unlinkByKeyIfFound() and Transients::unlinkByKeyIfFound() (for USE_HTCP off scenario). These were our 'innovations' to optimize 'heavy' StoreEntry loading. I.e., instead of StoreEntry::release() and eventually Transients::abandon() now we have unlinkByKeyIfFound() and no abandon(). So other worker clients keep 'hanging' on detached entry which became deleted.
To fix this, we can either:
* revert our changes in purgeEntriesByUrl() and load StoreEntry or
* teach Transients::unlinkByKeyIfFound() to do CollapsedForwarding::Broadcast(). There should be added a suitable api for this though.

What do you think?

Reply:
> revert our changes in purgeEntriesByUrl() and load StoreEntry

IIRC, loading StoreEntry is not an option in general because Controller::find() returns nil for not-yet-readable entries. See your PR description near "mark being-created
entries". Our changes were not simply an optimization. They addressed a problem with marking unreadable entries for deletion. We need to fix/improve our changes, not revert them.

> teach Transients::unlinkByKeyIfFound() to do CollapsedForwarding::Broadcast().

unlinkByKeyIfFound() or a related method should broadcast (in some cases): A worker must broadcast whenever it changes the state of a shared being-read entry -- the broadcast is the mechanism for readers in other workers to learn about those changes. Marking a transient entry for deletion requires broadcasting (when there are other entry readers).

I am not sure this broadcast should be initiated by the Transients::unlinkByKeyIfFound() method, by the corresponding Controller method, or by similar methods in the cache Stores. I hope you can find and defend the answer to that question. Ideally, there should be at most one broadcast per change. See below for more clues.


> There should be added a suitable api for this though.

I would check whether calling Transients::get() and broadcasting the resulting entry (or not broadcasting at all if it returns nil) is sufficient (and a good idea) in this case. That would still be different than relying on Controller::find() for deletion. Here, we are relying on (fast) get() to broadcast the change, _not_ to mark the store entries for deletion.

Note that, AFAICT, we do not need to broadcast if the transients entry has already been marked for deletion (which means we changed nothing) or when the transients entry is locked for writing (which means it cannot have any readers yet).

Also, if there is a local StoreEntry object already, we should probably use that instead of calling Transients::get(). Perhaps the code already has this optimization.

Combining all of the above, you might arrive at something like this (entry locking/unlocking not shown and may be missing):

```
    if (auto entry = transients->get(key))
        entry->release(); // broadcasts to inform other readers if needed
    else
        Controller::unlinkByKeyIfFound(key); // does not broadcast but there are no readers
```

As always, please check my reasoning. The overarching "broadcast changes when there are readers" principle should be correct, but I could miss various nuances/details.

Reply:
> IIRC, loading StoreEntry is not an option in general because Controller::find() returns nil for not-yet-readable entries. See your PR description near "mark being-created
entries". Our changes were not simply an optimization. They addressed a problem with marking unreadable entries for deletion. We need to fix/improve our changes, not revert them.

To clarify: my (1) suggestion did not assume 100% reverting purgeEntriesByUrl(): unlinkByKeyIfFound() method should stay of course, to handle situation when it is already impossible/not_yet_possible to 'get' the entry from store but it is still possible to get it by the key. I suggested to keep getting entry when we can(as the old code did). This would be a simple but not a perfect/complete solution though.

Reply:
My sketch is similar to what you are suggesting then, except it does not pay the relatively high price of creating a new StoreEntry object when transients do not have a usable StoreEntry object already.

If the optimization in my sketch causes trouble, then reusing the old code (i.e., creating a new StoreEntry object from store metadata whenever possible) is fine. Just add a source code comment for the reader who might be tempted to optimize this code by using Transients::get() alone.

Reply:
>  when the transients entry is locked for writing (which means it cannot have any readers yet)

Note that this situation is hardly possible in practice, I think. The just created(empty) StoreEntry goes into Controller::allowCollapsing() and Transients::startWriting(). After copying into the shared memory, it is immediately switched into 'appending' state. I.e., it is locked for writing while copyToShm() copies few metadata fields into shared memory.

Reply:
> Alex: when the transients entry is locked for writing

> Eduard: this situation is hardly possible in practice

s/hardly possible/frequently occurring but usually very brief/

... which is exactly the kind of situation that triggers difficult-to-find bugs.

> it is locked for writing while copyToShm() copies few metadata fields into shared memory

Agreed. Such copying usually takes just a fraction of a millisecond. However, various factors may slow down system calls, _and_ the code should work correctly during any brief period of time. With thousands of new transactions arriving per second, the chance that our code correctness will be tested is small but significant.

Reply:
@rousskov is it possible that CERN  might be a candidate for testing specific things?
I don't remember the exactly the details but they have a "CernVM File System"(https://cernvm.cern.ch/portal/filesystem) which from a vague memory they use with squid in some way.

Reply:
>  However, various factors may slow down system calls

I even doubt that there are system calls in Transients::copyToShm(). Looks like only memory copying without allocating new memory.

Reply:
@elico, I am not sure what you mean by "candidate for testing" but

* Squid is being used on CERN-related projects today (and they do report problems and sponsor improvements)
* I am not aware of any software testing services offered by CERN
* This PR is not a good place to discuss general Squid QA issues!

Reply:
@rousskov OK.
But it's kind of related to the subject in my mind.
They have big systems which uses squid-cache, I was referring to:
`With thousands of new transactions arriving per second, the chance that our code correctness will be tested is small but significant.`
and I think that CERN is one good client and use case which might be able to benefit from a good solution for this specific issue.

Reply:
> I even doubt that there are system calls in Transients::copyToShm(). Looks like only memory copying without allocating new memory.

Perhaps I should not have said "system calls". I do not know whether any system calls occur today while we hold an exclusive write lock. However, there can be delays (and that is what important in this context!) and, besides, future code may gain system calls even if there are none today (there is no "no system calls here!" invariant to rely on).

copyToShm() is not the only code executed while the transients entry is locked for writing. To start with, atomic locking/unlocking itself is a non-trivial operation that may require bus synchronization and other expensive/slow actions.

AFAIK, copying to shared memory may introduce significant delays because the kernel may need to allocate the (real) memory before we can copy and/or the kernel may need to flush what we have copied to disk, potentially blocking the process. Again, none of this may trigger system calls but it may cause delays.

Reply:
> The overarching "broadcast changes when there are readers" principle should be correct

Looks like such wording misses a case for 'zero' readers, i.e., when there is only a writer in the 'appending' mode. Logically, DELETE should abort such still detached single client similarly as if there were many collapsed detached clients. Broadcasting has a protection for zero readers:

```
if (!e.hasTransients() || !Store::Root().transientReaders(e)) {
    debugs(17, 7, "nobody reads " << e);
    return;
}
```

However, your sketch should work, because we create a reader in Transients::get():

```
if (auto entry = transients->get(key))
    entry->release(); // broadcasts to inform other readers or writer if needed
[...]
```

Reply:
> Alex: broadcast changes when there are readers

> Eduard: Looks like such wording misses a case for 'zero' readers, i.e., when there is only a writer in the 'appending' mode.

The case of zero/no readers is covered well AFAICT: There is no reason to broadcast unless there is at least one reader. Please note that, in this context, a "reader" is the Transients entry reader, not cache dir or memory store reader. In SMP environment, all collapsed slaves are readers.

> DELETE should abort such still detached single client

DELETE must not abort the writer. There is nothing wrong with the writing Store client. The corresponding HTTP client should get the full response. DELETE should only affect clients that come later (i.e., after DELETE).

> similarly as if there were many collapsed detached clients.

Unattached clients are somewhat special:

* In the ideal world, we would want them to (eventually) attach and receive the store entry because they came _before_ DELETE and, hence, should not be affected by it.
* However, the current Squid code will not let a client to attach to the marked-for-deletion store entry. This is a limitation of the current code/APIs, but we are not going to improve it during this PR (too difficult and mostly out of scope). Due to this limitation, when a (transient) entry is marked for deletion, we must notify all collapsed unattached readers so that they stop waiting -- they will never become attached to a disk or memory store.

The "broadcast changes when there are readers" rule is essential to deal with the current code limitations. The rule will not hurt much even if we improve the code to allow post-DELETE attachment, but it will result in unnecessary notifications in reaction to DELETE, so it should then be adjusted to exclude entry marking from the applicable "changes".

Reply:
> The case of zero/no readers is covered well AFAICT: There is no reason to broadcast unless there is at least one reader.

Ok, then that sketch is flawed or incomplete: we can do wrong broadcasting when there have been no readers before we called Transients::get(), thus creating a 'fake' reader in order to do that broadcasting. Also, IIRC, syncCollapsed() does not distinguish 'writer' or 'reader', i.e., it will abort detached collapsed writer as well.
To address the first problem we need one more check before doing Transients::get():

```
if (Store::Root().transientReaders(key))  {
    if (auto entry = transients->get(key))
        entry->release(); // broadcasts to inform other readers if needed
}
```

There is no Store::Root().transientReaders(key) for now: we would need to overload the existing method in this case.
For the second problem, there should be a way to distinguish transients readers from writer inside syncCollapsed() in order to skip aborting for detached writer.

Reply:
> we can do wrong broadcasting when there have been no readers before we called Transients::get()

If by "wrong" you mean excessive or unnecessary broadcasting, then I agree. If it is easy to avoid or reduce excessive broadcasting, please do that. Otherwise, I would not worry about this minor problem but add a comment to document the creation of an extra reader.

> there should be a way to distinguish transients readers from writer inside syncCollapsed() in order to skip aborting for detached writer

AFAICT, `mem_obj->smpCollapsed` has to be false for writers.

Please keep in mind that `syncCollapsed()` was originally written to update readers because only writers could generate updates, and `Broadcast()` does not send notifications to the _source_ of the update. With this PR, `syncCollapsed()` will be called for both readers and writers because DELETE may be received by a _reading_ worker, and `Broadcast()` does not know who the writer is.

You will need to adjust the code so that writers can handle notifications. AFAICT, they should not assert `smpCollapsed` like they do now, and they need to `release(true)` their entry (without aborting it).

Reply:
As we discussed, release() sometimes did not cause broadcasting in my tests (including the old code). I see the reason in Store::Controller::unlink(): it does not care about Transients, calling transientsAbandon().

Reply:
I agree that the lack of a `transients->unlink()` call in `Controller::unlink()` looks wrong, especially since `Controller::markForUnlink()` has the expected `transients->markForUnlink()` call.

However, I cannot say, without doing research, whether changing that method alone would guarantee broadcasts during `release()`. Please make sure `release()` broadcasts (directly or indirectly) whenever the state changes in such a way that a broadcast is required, regardless of whether `Controller::unlink()` has been called. Locating the best place(s) to (check state and) initiate that broadcast is not easy.

Reply:
> However, I cannot say, without doing research, whether changing that method alone would guarantee broadcasts during release()

No, it is not guaranteed: there is another problem. Transients::unlink() does broadcasting only for 'writers', i.e., when "e.mem_obj->xitTable.io == MemObject::ioWriting". That was the initial design: only writers can broadcast, but now, evidently, we need to broadcast from any worker, which received DELETE.

Reply:
> we need to broadcast from any worker, which received DELETE

To be more precise: 

* from any worker which changed the shared entry state
* to any worker reading that shared entry

This PR focus is on DELETE-triggered changes, and we can broaden these narrow rules if needed to ease implementation (e.g., broadcast to _all_ workers), but we should keep these rules in mind.

Please note that the rules have not changed. What changed is our understanding of their impact: We now know that DELETEs (and such) may require initiating broadcasts from non-writers.

Reply:
The related fixes for this thread discussion at 2765f0b.